src/backend/networking/rpc_utils.cpp

```diff
@@ -0,0 +1,252 @@
+            attrs[it]->attinhcount = tuple_desc_msg.attrs(it).attinhcount();
+            attrs[it]->attisdropped = tuple_desc_msg.attrs(it).attisdropped();
+            attrs[it]->attislocal = tuple_desc_msg.attrs(it).attislocal();
+            // It is int16 in postgres, so we convert it
+            attrs[it]->attlen = assert_range_cast_same<int16, int32>(
+                    tuple_desc_msg.attrs(it).attlen());
+            std::string attname = tuple_desc_msg.attrs(it).attname();
+            attname.copy(attrs[it]->attname.data,NAMEDATALEN);
+            attrs[it]->attndims = tuple_desc_msg.attrs(it).attndims();
+            attrs[it]->attnotnull = tuple_desc_msg.attrs(it).attnotnull();
+            // It is int16 in postgres, so we convert it
+            attrs[it]->attnum = assert_range_cast_same<int16, int32>(
+                    tuple_desc_msg.attrs(it).attnum());
+            attrs[it]->attrelid = tuple_desc_msg.attrs(it).attrelid();
+            attrs[it]->attstattarget = tuple_desc_msg.attrs(it).attstattarget();
+            // It is char in postgres, so we convert it
+            attrs[it]->attstorage = assert_range_cast_same<int8, int32>(
+                    tuple_desc_msg.attrs(it).attstorage());
+            attrs[it]->atttypid = tuple_desc_msg.attrs(it).atttypid();
+            attrs[it]->atttypmod = tuple_desc_msg.attrs(it).atttypmod();
+
+        }
+
+        tuple_desc->attrs = ppattrs;
+    } else {
+        tuple_desc->attrs = NULL;
+    }
+
+    //===----------------------------------------------------------------------===//
+    //    Set the basic value type
+    //===----------------------------------------------------------------------===//
+    tuple_desc->natts = attrs_count;
+    tuple_desc->tdhasoid = tuple_desc_msg.tdhasoid();
+    tuple_desc->tdrefcount = tuple_desc_msg.tdrefcount();
+    tuple_desc->tdtypeid = tuple_desc_msg.tdtypeid();
+    tuple_desc->tdtypmod = tuple_desc_msg.tdtypmod();
+
+    //===----------------------------------------------------------------------===//
+    //   Parse and create TupeDescCons structure
+    //===----------------------------------------------------------------------===//
+    // Note the three char* () is ended with \0 in postgres, such as appendStringInfoChar.
+    // So they are safe to be set in protobuf (as bytes) with on "size" parameter
+    // But when convert them back to char* we should set \0 at the end of char*
+    if (tuple_desc_msg.has_constr()) {
+        AttrDefault* attrdef = (AttrDefault*)malloc(sizeof(AttrDefault));
+        ConstrCheck* constrch = (ConstrCheck*)malloc(sizeof(ConstrCheck));
+        TupleConstr* tuple_constr = (TupleConstr*)malloc(sizeof(TupleConstr));
+
+        std::string adbinss = tuple_desc_msg.constr().adbin();
+        std::string ccbinss = tuple_desc_msg.constr().ccbin();
+        std::string ccnamess = tuple_desc_msg.constr().ccname();
+
+        char* adbin = (char*)malloc( adbinss.size() + 1 );
```

Comment:
That's a good suggestion. I also plan to use smart pointer here. But these raw pointers are inside tupleDesc which is a Postgres structure. We can't use std::move to transfer the ownership to these raw pointers. If we just use .get() to pass the raw pointer, we will get dangling pointers after the function call. 

Do you have any idea about that?